app/src/main/java/com/github/raulccabreu/redukt/middlewares/BaseAnnotatedMiddleware.kt

```diff
@@ -9,51 +9,82 @@ abstract class BaseAnnotatedMiddleware<T> : Middleware<T> {
+                    it.isAnnotationPresent(AfterAction::class.java) ||
+                    it.isAnnotationPresent(BeforeActions::class.java) ||
+                    it.isAnnotationPresent(AfterActions::class.java)
                 }.forEach {
-                    when {
-                        it.isAnnotationPresent(BeforeAction::class.java) -> addBeforeReduce(it)
-                        it.isAnnotationPresent(AfterAction::class.java) -> addAfterReduce(it)
-                    }
+                    if (it.isAnnotationPresent(BeforeAction::class.java))
+                        addBeforeMiddleware(it)
+                    if (it.isAnnotationPresent(AfterAction::class.java))
+                        addAfterMiddleware(it)
+                    if (it.isAnnotationPresent(BeforeActions::class.java))
+                        addBeforeActions(it)
+                    if (it.isAnnotationPresent(AfterActions::class.java))
+                        addAfterActions(it)
                 }
     }
 
-    private fun addBeforeReduce(method: Method) {
+    private fun addBeforeMiddleware(method: Method) {
         val annotation = method.getAnnotation(BeforeAction::class.java) as BeforeAction
 
-        if (annotation.action.isBlank())
-            throw IllegalArgumentException("BeforeReduce action cannot be empty")
-
-        if (method.parameterTypes.size != 2)
-            throw InvalidParameterException(
-                    "The method ${method.name} must accept: State and Action")
+        verifyActionIsBlank (annotation.action)
+        verifyNumberOfMethods(method)
 
         befores.put(annotation.action, method)
     }
 
-    private fun addAfterReduce(method: Method) {
+    private fun addAfterMiddleware(method: Method) {
         val annotation = method.getAnnotation(AfterAction::class.java) as AfterAction
 
-        if (annotation.action.isBlank())
-            throw IllegalArgumentException("BeforeReduce action cannot be empty")
+        verifyActionIsBlank (annotation.action)
+        verifyNumberOfMethods(method)
+
+        afters.put(annotation.action, method)
+    }
+
+    private fun addBeforeActions(method: Method) {
+        verifyNumberOfMethods(method)
+
+        interceptBefores.add(method)
+    }
+
+    private fun addAfterActions(method: Method) {
+        verifyNumberOfMethods(method)
+
+        interceptAfters.add(method)
+    }
+
+    private fun verifyActionIsBlank(action: String) {
+        if (action.isBlank())
+            throw IllegalArgumentException("Action cannot be empty")
+    }
 
+    private fun verifyNumberOfMethods(method: Method) {
         if (method.parameterTypes.size != 2)
             throw InvalidParameterException(
                     "The method ${method.name} must accept: State and Action")
-
-        afters.put(annotation.action, method)
     }
 
     override fun before(state: T, action: Action<*>) {
+        if (!canExecute(state)) return
```

Comment:
We should make before and after a final method, because if the user overrides this method, the base annotated middleware will not works properly.

Reply:
I think the "fun" is final by default. If you enable to change, you need add "open fun". But I'm not sure.

Reply:
https://kotlinlang.org/docs/reference/classes.html#overriding-methods

Reply:
Hmm! Same doubt here! Looks that in the interface it doesn't have open, but maybe it is an exception for interfaces.

![image](https://user-images.githubusercontent.com/2291529/34012622-4bfbb392-e0f3-11e7-833c-eb8acb44f6d4.png)

Reply:
Hmm! I check here and the documentation says that override methods are open by default:

![image](https://user-images.githubusercontent.com/2291529/34012655-6b69bc42-e0f3-11e7-932b-cea40a99e521.png)

Reply:
This is true, because it is override methods